### PR TITLE
refactor allocation spec creation

### PR DESCRIPTION
**Pull request by majst01**
With this PR the duplicated code to fill the internal struct allocationspec was refactored to be reusable for machine and firewall allocation. Also the FirewallCreateRequest.HA field was never used and deprecated here.

This PR is a preparation for #232 to make the review of this PR easier.

### CODE REVIEW

File: cmd/metal-api/internal/service/machine-service.go [hunk 1]
```diff
@@ -942,54 +967,58 @@ func (r machineResource) allocateMachine(request *restful.Request, response *res
 	if requestPayload.UserData != nil {
 		userdata = *requestPayload.UserData
 	}
+	if requestPayload.Networks != nil && len(requestPayload.Networks) <= 0 {
```
Comment (majst01) [inline cmd/metal-api/internal/service/machine-service.go:970]: We should check for nil networks as well !

File: cmd/metal-api/internal/service/machine-service.go [hunk 2]
```diff
@@ -942,54 +967,58 @@ func (r machineResource) allocateMachine(request *restful.Request, response *res
...
-		Role:               metal.RoleMachine,
-	}
-
-	m, err := allocateMachine(utils.Logger(request).Sugar(), r.ds, r.ipamer, &spec, r.mdc, r.actor, r.grpcServer)
-	if checkError(request, response, utils.CurrentFuncName(), err) {
-		utils.Logger(request).Sugar().Errorw("machine allocation went wrong", "error", err)
-		return
+	partitionID := requestPayload.PartitionID
+	sizeID := requestPayload.SizeID
+	// Allocation of a specific machine is requested, therefor size and partition are not given, fetch them
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go:990]:
```suggestion
	// Allocation of a specific machine is requested, therefore size and partition are not given, fetch them
```

File: cmd/metal-api/internal/service/machine-service.go [hunk 3]
```diff
@@ -942,54 +967,58 @@ func (r machineResource) allocateMachine(request *restful.Request, response *res
...
+	if uuid != "" {
+		m, err := ds.FindMachineByID(uuid)
+		if err != nil {
+			return nil, fmt.Errorf("uuid given but no machine found with uuid:%s err:%w", uuid, err)
+		}
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/machine-service.go:995]: Introduces an additional backend call, which I think gets repeated in `findMachineCandidate`. Maybe it makes sense to store the pointer to this machine entity in the allocation spec and we can get rid off UUID plus the additional backend call.

**Review by Gerrit91 — COMMENTED** [review]
Thanks for putting this into a separate PR. Makes it much better to understand. :)

File: cmd/metal-api/internal/service/machine-service.go (hunk 3 above)
Comment (majst01) [reply cmd/metal-api/internal/service/machine-service.go:996]: Good idea, done

**Comment by majst01** [comment]
Unrelated but its possible to allocate a firewall which leads to a failure during installation:

```
{"level":"panic","ts":1637767309.1289444,"caller":"cmd/root.go:93","msg":"expectation at least one public network (private: false, underlay: false) is present failed","stacktrace":"github.com/metal-stack/metal-networker/cmd.configure\n\tgithu
b.com/metal-stack/metal-networker/cmd/root.go:93\ngithub.com/metal-stack/metal-networker/cmd.glob..func1\n\tgithub.com/metal-stack/metal-networker/cmd/firewall.go:24\ngithub.com/spf13/cobra.(*Command).execute\n\tgithub.com/spf13/cobra@v1.1.3/
command.go:852\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tgithub.com/spf13/cobra@v1.1.3/command.go:960\ngithub.com/spf13/cobra.(*Command).Execute\n\tgithub.com/spf13/cobra@v1.1.3/command.go:897\ngithub.com/metal-stack/metal-networker/cmd.
Execute\n\tgithub.com/metal-stack/metal-networker/cmd/root.go:43\nmain.main\n\tgithub.com/metal-stack/metal-networker/main.go:6\nruntime.main\n\truntime/proc.go:225"}                                                                            
panic: expectation at least one public network (private: false, underlay: false) is present failed  
```

**Review by Gerrit91 — APPROVED** [review]

**Pull request merged**